### PR TITLE
[bugfix] fix particle index initialization in gadget_fof frontend

### PR DESCRIPTION
**Pull request by brittonsmith**
## PR Summary

The `gadget_fof` frontend was giving the following error when loading data:
```
Traceback (most recent call last):
  File "test.py", line 4, in <module>
    ds.field_list
  File "/Users/britton/Documents/work/yt/yt-4.0/yt/data_objects/static_output.py", line 556, in field_list
    return self.index.field_list
  File "/Users/britton/Documents/work/yt/yt-4.0/yt/data_objects/static_output.py", line 514, in index
    self, dataset_type=self.dataset_type)
  File "/Users/britton/Documents/work/yt/yt-4.0/yt/frontends/gadget_fof/data_structures.py", line 48, in __init__
    super(GadgetFOFParticleIndex, self).__init__(ds, dataset_type)
  File "/Users/britton/Documents/work/yt/yt-4.0/yt/geometry/particle_geometry_handler.py", line 42, in __init__
    super(ParticleIndex, self).__init__(ds, dataset_type)
  File "/Users/britton/Documents/work/yt/yt-4.0/yt/geometry/geometry_handler.py", line 49, in __init__
    self._setup_geometry()
  File "/Users/britton/Documents/work/yt/yt-4.0/yt/frontends/gadget_fof/data_structures.py", line 115, in _setup_geometry
    self._calculate_particle_count()
  File "/Users/britton/Documents/work/yt/yt-4.0/yt/frontends/gadget_fof/data_structures.py", line 56, in _calculate_particle_count
    for ptype in self.ds.particle_types_raw])
  File "/Users/britton/Documents/work/yt/yt-4.0/yt/frontends/gadget_fof/data_structures.py", line 56, in <listcomp>
    for ptype in self.ds.particle_types_raw])
AttributeError: 'GadgetFOFParticleIndex' object has no attribute 'data_files'
```

The order of initialization in `ParticleIndex` wasn't quite compatible with the frontend. Additionally, it looks like some io chunking was implemented and so this adds support for that as well.

### CODE REVIEW

**Comment by ngoldbaum** [comment]
Cool, thanks for this! I guess I had unintentionally broken Gadget-FoF at some point, sorry for the trouble.

**Comment by brittonsmith** [comment]
No problemo!

**Comment by brittonsmith** [comment]
It looks like I need to thread the chunking io through the io routines as well. I will do that.

**Comment by brittonsmith** [comment]
I've threaded the chunked io through the rest of the halo catalog-style frontends. Let me know if I've missed something or this was unnecessary. I've included a backport of PR #2139 when I was thinking about indexing hdf5 reads directly, but I didn't end up doing that, so I can remove that change if preferred.

**Comment by ngoldbaum** [comment]
It's fine to include that. I'm gonna try to merge yt-4.0 with master this week so we don't need to worry about that at the workshop.

File: doc/source/cookbook/extract_fixed_resolution_data.py [hunk 1]
```diff
@@ -30,4 +30,4 @@
 
 # If we want to then access this datacube in the h5 file, we can now...
 f = h5py.File("my_data.h5", "r")
-print(f["density"].value)
+print(f["density"][()])
```
Comment (MatthewTurk) [inline doc/source/cookbook/extract_fixed_resolution_data.py]: I know this is how it gets done nowadays, but I think it might be more in keeping with the style if we do a colon.

File: yt/frontends/chombo/data_structures.py [hunk 2]
```diff
@@ -188,9 +188,9 @@ def _parse_index(self):
         for lev_index, lev in enumerate(self._levels):
             level_number = int(re.match('level_(\d+)',lev).groups()[0])
             try:
-                boxes = f[lev]['boxes'].value
+                boxes = f[lev]['boxes'][()]
```
Comment (MatthewTurk) [inline yt/frontends/chombo/data_structures.py]: same here and below

File: yt/frontends/gadget_fof/data_structures.py [hunk 3]
```diff
@@ -45,7 +45,13 @@
 
 class GadgetFOFParticleIndex(ParticleIndex):
     def __init__(self, ds, dataset_type):
-        super(GadgetFOFParticleIndex, self).__init__(ds, dataset_type)
+        self.dataset_type = dataset_type
+        self.dataset = weakref.proxy(ds)
+        self.float_type = np.float64
+        self._setup_data_io()
+        self._setup_filenames()
+        super(ParticleIndex, self).__init__(ds, dataset_type)
```
Comment (MatthewTurk) [inline yt/frontends/gadget_fof/data_structures.py]: Can you explain why the change in the super call?  I'm not sure I follow.

**Comment by ngoldbaum** [comment]
@brittonsmith agree with matt about the super calls, I think there's a mistake there.

File: doc/source/cookbook/extract_fixed_resolution_data.py (hunk 1 above)
Comment (brittonsmith) [reply doc/source/cookbook/extract_fixed_resolution_data.py]: This is the recommended syntax from `h5py` that was included in the deprecation warning for `.value`.

File: yt/frontends/gadget_fof/data_structures.py (hunk 3 above)
Comment (brittonsmith) [reply yt/frontends/gadget_fof/data_structures.py]: Oh yeah, this was so I could change the ordering of some things that happen in `__init__`. Specifically, I think I needed `_setup_data_io` and `_setup_filenames` to happen before things that happen in `Index.__init__`. Much of this change is taking what happens in `ParticleIndex.__init__` and moving it here. That's why I don't switched from calling `super(GadgetFOFParticleIndex, ...)` to `super(ParticleIndex, ...)`.

File: doc/source/cookbook/extract_fixed_resolution_data.py (hunk 1 above)
Comment (brittonsmith) [reply doc/source/cookbook/extract_fixed_resolution_data.py]: I should also say that these changes are just cherry-picked from a PR already merged into master.

**Comment by brittonsmith** [comment]
This latest change of reordering the `__init__` of `ParticleIndex` appears to have broken the `stream` frontend. I'd like to return to the original method. If I stick some comments in there, will that suffice for now?

**Comment by ngoldbaum** [comment]
Would you mind if I poke at it?

File: doc/source/cookbook/extract_fixed_resolution_data.py (hunk 1 above)
Comment (MatthewTurk) [reply doc/source/cookbook/extract_fixed_resolution_data.py]: OK, comment retracted!
